collections/pair.go

```diff
@@ -0,0 +1,231 @@
+		written, err := p.keyCodec1.EncodeNonTerminal(buffer, *pair.key1)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	if pair.key2 != nil {
+		written, err := p.keyCodec2.Encode(buffer[writtenTotal:], *pair.key2)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	return writtenTotal, nil
+}
+
+func (p pairKeyCodec[K1, K2]) Decode(buffer []byte) (int, Pair[K1, K2], error) {
+	readTotal := 0
+	read, key1, err := p.keyCodec1.DecodeNonTerminal(buffer)
+	if err != nil {
+		return 0, Pair[K1, K2]{}, err
+	}
+	readTotal += read
+	read, key2, err := p.keyCodec2.Decode(buffer[read:])
+	if err != nil {
+		return 0, Pair[K1, K2]{}, err
+	}
+
+	readTotal += read
+	return readTotal, Join(key1, key2), nil
+}
+
+func (p pairKeyCodec[K1, K2]) Size(key Pair[K1, K2]) int {
+	size := 0
+	if key.key1 != nil {
+		size += p.keyCodec1.SizeNonTerminal(*key.key1)
+	}
+	if key.key2 != nil {
+		size += p.keyCodec2.Size(*key.key2)
+	}
+	return size
+}
+
+func (p pairKeyCodec[K1, K2]) Stringify(key Pair[K1, K2]) string {
+	b := new(strings.Builder)
+	b.WriteByte('(')
+	if key.key1 != nil {
+		b.WriteByte('"')
+		b.WriteString(p.keyCodec1.Stringify(*key.key1))
+		b.WriteByte('"')
+	} else {
+		b.WriteString("<nil>")
+	}
+	b.WriteString(", ")
+	if key.key2 != nil {
+		b.WriteByte('"')
+		b.WriteString(p.keyCodec2.Stringify(*key.key2))
+		b.WriteByte('"')
+	} else {
+		b.WriteString("<nil>")
+	}
+	b.WriteByte(')')
+	return b.String()
+}
+
+func (p pairKeyCodec[K1, K2]) KeyType() string {
+	return fmt.Sprintf("Pair[%s, %s]", p.keyCodec1.KeyType(), p.keyCodec2.KeyType())
+}
+
+func (p pairKeyCodec[K1, K2]) EncodeNonTerminal(buffer []byte, pair Pair[K1, K2]) (int, error) {
+	writtenTotal := 0
+	if pair.key1 != nil {
+		written, err := p.keyCodec1.EncodeNonTerminal(buffer, *pair.key1)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	if pair.key2 != nil {
+		written, err := p.keyCodec2.EncodeNonTerminal(buffer[writtenTotal:], *pair.key2)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	return writtenTotal, nil
+}
+
+func (p pairKeyCodec[K1, K2]) DecodeNonTerminal(buffer []byte) (int, Pair[K1, K2], error) {
+	readTotal := 0
+	read, key1, err := p.keyCodec1.DecodeNonTerminal(buffer)
+	if err != nil {
+		return 0, Pair[K1, K2]{}, err
+	}
+	readTotal += read
+	read, key2, err := p.keyCodec2.DecodeNonTerminal(buffer[read:])
+	if err != nil {
+		return 0, Pair[K1, K2]{}, err
+	}
+
+	readTotal += read
+	return readTotal, Join(key1, key2), nil
+}
+
+func (p pairKeyCodec[K1, K2]) SizeNonTerminal(key Pair[K1, K2]) int {
+	size := 0
+	if key.key1 != nil {
+		size += p.keyCodec1.SizeNonTerminal(*key.key1)
+	}
+	if key.key2 != nil {
+		size += p.keyCodec2.SizeNonTerminal(*key.key2)
+	}
+	return size
+}
+
+// NewPairRange creates a new PairRange which will prefix over all the keys
+// starting with the provided prefix.
+func NewPairRange[K1, K2 any](prefix K1) *PairRange[K1, K2] {
```

Comment:
How about a second constructor for K1 and K2 as prefix?

Reply:
I'm scared that prefixing over K1 and K2 is misleading as it would work only for let's say any K1 but only certain K2, it would work for if K2 is string/bytes, but not if K2 is uint64 (fixed size).

Reply:
I could see value though in iterating between some `Join(K1, K2)`up till `Join(K1,K2)`, but for that use-case maybe the simple `Range` is better than `PairRange`.

Reply:
> I'm scared that prefixing over K1 and K2 is misleading as it would work only for let's say any K1 but only certain K2, it would work for if K2 is string/bytes, but not if K2 is uint64 (fixed size).

Why will it not work for `uint64`? It should be fine